tools/pybug.py

```diff
@@ -0,0 +1,1544 @@
+        cmd_srom_type,
+        "25aa1024|25aa080a|25aa160b",
+        "Set SROM type"),
+    "srom_dump": (
+        cmd_srom_dump,
+        "<file.F> <addr.X> <len.D>",
+        "Dump serial ROM data"),
+    "srom_write": (
+        cmd_srom_write,
+        "<file.F> <addr.X>",
+        "Write serial ROM data"),
+    "srom_erase": (
+        cmd_srom_erase,
+        "",
+        "Erase (all) serial ROM data"),
+    "srom_init": (
+        cmd_srom_init,
+        "<Flag.X> <MAC.M> <ip_addr.P> <gw_addr.P> <net_mask.P> <port.D>",
+        "Initialise serial ROM"),
+    "remap": (
+        cmd_remap,
+        "<core.D> [phys|virt]",
+        "Remove bad core from core map"),
+    "p2p_route": (
+        cmd_p2p_route,
+        "[on|off]",
+        "Control P2P routing"),
+    "app_dump": (
+        cmd_app_dump,
+        "",
+        "Show app data for this chip"),
+    "cmd": (
+        cmd_cmd,
+        '<cmd.D> <arg1.X> <arg2.X> <arg3.X>',
+        'User specified command'),
+}
+
+# ------------------------------------------------------------------------------
+
+
+def usage():
+    print("usage: pybug <options> <hostname>", file=sys.stderr)
+    print("  -bmp  <name>[/<slots>]   - set BMP target", file=sys.stderr)
+    print("  -version                 - print version number", file=sys.stderr)
+    print("  -norl                    - don't use 'ReadLine'", file=sys.stderr)
+    print("  -expert                  - set 'expert' mode", file=sys.stderr)
+    print("  -debug <value>           - set debug variable", file=sys.stderr)
+    sys.exit(1)
+
+
+def process_args():
+    global spinn_target, bmp_target, expert, _readline, debug
+    global bmp_range
+    _range = "0"
+    args = list(sys.argv)
+    while args:
+        arg = args.pop(0)
+        if arg == "-bmp":
+            bmp_target = args.pop(0)
+            m = re.match(r"^(.*)/(\S+)$", bmp_target)
+            if m:
+                bmp_target, _range = m.groups()
+        elif arg == "-version":
+            cmd_version(None)
+            sys.exit()
+        elif arg == "-debug":
+            debug = int(args.pop(0))
+        elif arg == "-norl":
+            _readline = False
+        elif arg == "-expert":
+            expert = True
+        elif not re.match(r"^-", arg):
+            spinn_target = arg
+            break
+        else:
+            usage()
+    if args:
+        usage()
+
+    bmp_range = parse_bits(_range, 0, 23)
+    if spinn_target is None:
+        print("target not specified", file=sys.stderr)
+        sys.exit(1)
+    prompt = spinn_target
+    if not re.match(r"^\d", prompt):
+        prompt = re.sub(r"\..+", "", prompt)
+    prompt += ":0,0,0 > "
+    return prompt
+
+
+def open_targets():
+    global spin, sv, bmp
+    spin = SCAMPCmd(target=spinn_target, port=spin_port, debug=debug)
+    sv = Struct(scp=spin)
+    if bmp_target is not None:
+        bmp = BMPCmd(target=bmp_target, port=bmp_port, debug=debug)
+
+
+class Completer(object):
```

Comment:
Inner class?

1. Does it need to be an inner Class
2. Use a underscore name.

Reply:
> Use a underscore name.

This.